tests/unit/package_managers/yarn/test_resolver.py

```diff
@@ -804,3 +865,103 @@ def test_create_components_cache_path_reported_but_missing(rooted_tmp_path: Root
+        # HttpsLocators
+        (
+            _ResolvedPackage(
+                locator=HttpsLocator(url="https://example.com/foo.tgz"),
+                name="https-locator-package",
+                version="2.0.0",
+                checksum="abcdef",
+            ),
+            "pkg:npm/https-locator-package@2.0.0?checksum=sha512:abcdef&download_url=https://example.com/foo.tgz",
+        ),
+        # WorkspaceLocators
+        (
+            _ResolvedPackage(
+                locator=WorkspaceLocator(scope=None, name="", relpath=Path("path")),
+                name="workspace-locator-package",
+                version="3.0.0",
+                checksum=None,
+            ),
+            f"pkg:npm/workspace-locator-package@3.0.0?vcs_url={MOCK_REPO_VCS_URL}#path",
+        ),
+        # FileLocators
+        (
+            _ResolvedPackage(
+                locator=FileLocator(
+                    relpath=Path("external-packages/xyz-4.0.0.tgz"),
+                    locator=WorkspaceLocator(scope=None, name="", relpath=Path("some-path")),
+                ),
+                name="file-locator-package",
+                version="4.0.0",
+                checksum=None,
+            ),
+            f"pkg:npm/file-locator-package@4.0.0?vcs_url={MOCK_REPO_VCS_URL}#some-path/external-packages/xyz-4.0.0.tgz",
+        ),
+        # LinkLocators
+        (
+            _ResolvedPackage(
+                locator=LinkLocator(
+                    scope=None,
+                    name="",
+                    relpath=Path("external-packages/strip-ansi-5.0.0.tgz"),
+                    locator=WorkspaceLocator(scope=None, name="", relpath=Path("path")),
+                ),
+                name="link-locator-package",
+                version="5.0.0",
+                checksum=None,
+            ),
+            f"pkg:npm/link-locator-package@5.0.0?vcs_url={MOCK_REPO_VCS_URL}#path/external-packages/strip-ansi-5.0.0.tgz",
+        ),
+        # PortalLocators
+        (
+            _ResolvedPackage(
+                locator=PortalLocator(
+                    relpath=Path("external-packages/holy-hand-grenade"),
+                    locator=WorkspaceLocator(scope=None, name="", relpath=Path("some-path")),
+                ),
+                name="portal-locator-package",
+                version="6.0.0",
+                checksum=None,
+            ),
+            f"pkg:npm/portal-locator-package@6.0.0?vcs_url={MOCK_REPO_VCS_URL}#some-path/external-packages/holy-hand-grenade",
+        ),
+        # PatchLocators
+        (
+            _ResolvedPackage(
+                locator=PatchLocator(
+                    package=NpmLocator(scope=None, name="", version=""), patches=[], locator=None
+                ),
+                name="patch-locator-package",
+                version="7.0.0",
+                checksum=None,
+            ),
+            "pkg:npm/patch-locator-package@7.0.0",
+        ),
+    ],
+)
+def test_generate_purl_for_package(
```

Comment:
I wonder if adding this test here brings any additional value compared to the already existing `test_create_components_single_package` and `test_create_components_patched_packages`, since `_generate_purl_for_package` is called as part of the component creation. Does this bring any additional coverage?

Reply:
No, the coverage is the same. I can leave it there or drop it, does not matter to me. WDYT @everyone

Reply:
I will drop it since I move the **_generate_purl_for_package** function to be a method of _**ComponentResolver**

Reply:
> test_create_components_single_package and test_create_components_patched_packages

While the coverage may be the same I wonder if we could potentially decrease the scope of those tests? I'm asking because I think **test_generate_purl_for_package** is the correct test for this implementation, the others  are transitive only which may not be obvious at first glance and one has to go digging what happens underneath.

Reply:
I think this test's scope is given by the number of different "locator strings vs expected purls" scenarios. I think the currents tests have a good coverage on that.

I do agree that it's less clear to have this specific piece tested as part of `create_components` instead of only `_generate_purl_for_package`, but at the same time, I wonder if it's worth duplicating the scenarios to have a separate test, that's why I raised this point.